src/main/java/li/cil/oc2/api/inet/LinkLocalLayerInternetProvider.java

```diff
@@ -0,0 +1,38 @@
+package li.cil.oc2.api.inet;
+
+import li.cil.oc2.common.inet.NullLinkLocalLayer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * An {@link InternetProvider} partial implementation that expects an {@link LinkLocalLayer} implementation from
+ * protected method {@link LinkLocalLayerInternetProvider#provideLinkLocalLayer()}.
+ *
+ * @see InternetProvider
+ * @see LinkLocalLayer
+ */
+public abstract class LinkLocalLayerInternetProvider implements InternetProvider {
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    /**
+     * This method is called from {@link LinkLocalLayerInternetProvider#provideInternet()} in order to get an
+     * {@link LinkLocalLayer} implementation.
+     * If it fails, then a dummy {@link LinkLocalLayer} implementation will be used.
+     * Dummy implementation will do nothing.
+     *
+     * @return an implementation of link local TCP/IP layer for internet cards
+     * @throws Exception this method is allowed to fail with an exception
+     */
+    protected abstract LinkLocalLayer provideLinkLocalLayer() throws Exception;
+
+    @Override
+    public final LinkLocalLayer provideInternet() {
```

Comment:
It is assumed that implementations of `*LayerInternetProvider` are guaranteed to override the behavior of the Internet card at one of the layers of the network stack. The method `LinkLocalLayerInternetProvider::provideInternet` in this case does not use the default Internet card implementation, but an implementation that does nothing in case of an error (a stub). This is done in order to not crash the server and to make it easier to write the implementation of the `*LayerInternetProvider` classes. In addition, I thought about the cases when, in case of an error in the implementation of `SessionLayerInternetProvider`, the developer of the` SessionLayerProvider` implementation himself wants to return the stub implementation of `SessionLayer`. However, the `SessionLayerProvider` is too high in the TCP/IP stack and therefore the default implementations of all previous levels of the stack will perform meaningless operations on packets outgoing from the virtual machine.

Looking back, I think that it was completely unnecessary. I agree that `Optional<LinkLocalLayer>` will be better in this case. However, I now think it makes sense to leave the method signature as it is now, while still offering access to default implementations and stubs. For example, I can introduce the methods `NetworkLayerInternetProvider::defaultNetworkLayer` and `NetworkLayerInternetProvider::nullNetworkLayer`.

```java
public abstract class NetworkLayerInternetProvider extends LinkLocalLayer {
    protected static NetworkLayer defaultNetworkLayer() { ... }

    protected static NetworkLayer nullNetworkLayer() {
        return NullNetworkLayer.INSTANCE;
    }

    protected abstract NetworkLayer provideNetworkLayer();

    @Override
    public final LinkLocalLayer provideLinkLocalLayer() {
        final NetworkLayer networkLayer = provideNetworkLayer();
        if (networkLayer == NullNetworkLayer.INSTANCE) {
            // Optimize stabs
            return NullLinkLocalLayer.INSTANCE;
        } else {
            return new DefaultLinkLocalLayer(networkLayer);
        }
    }
}
```

A basic implementation of the new `NetworkLayerInternetProvider` class would look like this:

```java
public abstract class MyNetworkLayerInternetProvider extends NetworkLayerInternetProvider {
    @Override
    protected NetworkLayer provideNetworkLayer() {
        try {
            // Wrap the default implementation
            return new MyNetworkLayer(defaultLinkLocalLayer());
        } catch (Exception e) {
            // Error! Return stub. Do not crush.
            printException(e);
            return nullLinkLocalLayer();
        }
    }
}
```

If you are satisfied with this solution, then I can do it this way. If you find it too complicated for developers implementing the `*LayerInternetProvider`, then I will rewrite this part using `Optional` as you suggested.